### PR TITLE
[Routing] TurnDirections refactoring

### PR DESCRIPTION
**Pull request by AntonM030481**
Total refactoring of GetTurnDirection() and it's subfunctions.
A lot of new integration tests added (one additional map is required - Cyprus).
Old ones fixed where necessary.
Now only 3 tests failed:
bicycle_turn_test.cpp::RussiaMoscowSevTushinoParkBicycleWayTurnTest (arguable, see todo there)
turn_test.cpp::Russia_Moscow_BolshoyKislovskiyPerBolshayaNikitinskayaUl_TurnTest (bug of other system, see todo there)
turn_test.cpp::Netherlands_Barneveld_TurnTest (complicated new test).

Fixed:
#2417
#2438

Results of tests from pedestrian_directions.cpp are the same as before.

### CODE REVIEW

**Comment by AntonM030481** [comment]
I had to change struct TurnInfo reference members:
```
LoadedPathSegment const & m_ingoing;
LoadedPathSegment const & m_outgoing;
```
to pointers:
```
LoadedPathSegment const * m_ingoing;
LoadedPathSegment const * m_outgoing;
```
because of new function, which require assignment operator (incompatible with reference members).
```
TurnInfo turnInfo;
bool validTurnInfo = GetTurnInfo(result, outgoingSegmentIndex, vehicleSettings, turnInfo);
```
@biodranik Maybe better solution will be to use std::optional here and go back to references:
std::optional< turnInfo> = GetTurnInfo(result, outgoingSegmentIndex, vehicleSettings);

BTW, std::optional was introduced in C++17, but C++14 is used.
Frankly speaking I don't understand why it compiles now despite of -std=c++14 in CMAKE_CXX_FLAGS.

File: routing/turns_generator.cpp [hunk 1]
```diff
@@ -1062,238 +1014,242 @@ bool PathIsFakeLoop(std::vector<geometry::PointWithAltitude> const & path)
...
-
-  if (PathIsFakeLoop(segments[outgoingSegmentIndex].m_path))
-    return;
-
-  bool const isStartFakeLoop = PathIsFakeLoop(segments[outgoingSegmentIndex - 1].m_path);
-
-  if (isStartFakeLoop && outgoingSegmentIndex < 2)
+  TurnInfo turnInfo;
+  bool validTurnInfo = GetTurnInfo(result, outgoingSegmentIndex, vehicleSettings, turnInfo);
+  if (!validTurnInfo)
```
Comment (vng) [inline routing/turns_generator.cpp:1235]: if (!GetTurnInfo(...))

File: routing/turns_generator.cpp [hunk 2]
```diff
@@ -16,55 +16,62 @@
...
-  ftypes::HighwayClass m_smallestRouteRoadClass = ftypes::HighwayClass::Error;
-  // Let's consider all possible ways from the turn except for the way along the route and
-  // the way which let us to make a U-turn by going along the ingoing segment.
-  // |m_biggestPossibleTurnRoadClass| should be equal to the class of the biggest road of such roads.
-  ftypes::HighwayClass m_biggestPossibleTurnRoadClass = ftypes::HighwayClass::Error;
-};
-
-bool IsHighway(ftypes::HighwayClass hwClass, bool isLink)
+// Min difference of route and alternative turn abs angles in degrees
+// to ignore alternative alternative turn (when route direction is GoStraight).
```
[the remark this answers is not in the document]
Comment (AntonM030481) [reply routing/turns_generator.cpp:28]: ок

File: routing/turns_generator.cpp [hunk 3]
```diff
@@ -16,55 +16,62 @@
...
-/// and returns true. Or does not change |turn| and returns false.
-/// \note The function makes a decision about |turn| based on geometry of the route and turn
+/// \brief Return |CarDirection::ExitHighwayToRight| or |CarDirection::ExitHighwayToLeft|
+/// or return |CarDirection::None| if no exit is detected.
+/// \note The function makes a decision about turn based on geometry of the route and turn
 /// candidates, so it works correctly for both left and right hand traffic.
-bool IsExit(TurnCandidates const & possibleTurns, TurnInfo const & turnInfo,
-            Segment const & firstOutgoingSeg, CarDirection intermediateDirection, CarDirection & turn)
+CarDirection TryToGetExitDirection(TurnCandidates const & possibleTurns, TurnInfo const & turnInfo,
+            Segment const & firstOutgoingSeg, CarDirection const intermediateDirection)
```
[the remark this answers is not in the document]
Comment (AntonM030481) [reply routing/turns_generator.cpp:64]: ok

File: routing/turns_generator.cpp [hunk 4]
```diff
@@ -367,6 +328,23 @@ m2::PointD GetPointByIndex(TUnpackedPathSegments const & segments, RoutePointInd
...
+  switch (highwayClass)
+  {
+    case HighwayClass::Trunk:         speedMetersPerSecond = 100000.0/60/60; break;
+    case HighwayClass::Primary:       speedMetersPerSecond = 70000.0/60/60; break;
+    case HighwayClass::Secondary:     speedMetersPerSecond = 70000.0/60/60; break;
+    case HighwayClass::Tertiary:      speedMetersPerSecond = 50000.0/60/60; break;
+    case HighwayClass::LivingStreet:  speedMetersPerSecond = 20000.0/60/60; break;
+    case HighwayClass::Service:       speedMetersPerSecond = 10000.0/60/60; break;
+    case HighwayClass::Pedestrian:    speedMetersPerSecond = 50000.0/60/60; break;
+    default:                          speedMetersPerSecond = 500000.0/60/60; break;
```
[the remark this answers is not in the document]
Comment (AntonM030481) [reply routing/turns_generator.cpp:343]: Agree. Even better - found
measurement_utils::KmphToMps()
and used it.
BTW there is duplicating function transit::KmphToMps()

File: routing/turns_generator.cpp [hunk 5]
```diff
@@ -16,55 +16,62 @@
 using namespace routing;
 using namespace routing::turns;
 using namespace std;
+using namespace ftypes;
```
[the remark this answers is not in the document]
Comment (AntonM030481) [reply routing/turns_generator.cpp:19]: Do we need anonymous namespace?

File: routing/turns_generator.cpp (hunk 1 above)
Comment (AntonM030481) [reply routing/turns_generator.cpp:1235]: Can I use https://github.com/organicmaps/organicmaps/pull/2508#issuecomment-1117040390?

File: routing/turns_generator.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (vng) [reply routing/turns_generator.cpp:22]: keep

File: routing/turns_generator.cpp (hunk 1 above)
Comment (AntonM030481) [reply routing/turns_generator.cpp]: Changed to 
if (!GetTurnInfo(...))

**Comment by AntonM030481** [comment]
If we had enough tests for pedestrians i would change MakeTurnAnnotationPedestrian() too.
No it's almost full copy of MakeTurnAnnotation(). And both are not trivial.

File: routing/turns_generator.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (AntonM030481) [reply routing/turns_generator.cpp:22]: Is it OK?

[the remark this answers is not in the document]
Comment (vng) [reply routing/turns_generator.cpp:22]: ok

**Comment by vng** [comment]
@AntonM030481 I do merge? Or you gonna append the PR?

**Comment by AntonM030481** [comment]
@vng Sorry for huge amount of changes.
But I hope now it is more easy to understand and modify.
So next time I suppose changes will be not so dramatic.
Thank you for quick review.

**Pull request merged**